MM-36733 : Replace clickable divs with buttons for emoji skin tone picker

thinkGeist opened the pull request:
#### Summary


- Changed the `div` items with `onClick` methods to a `button` to allow the elements to be tabbed to
- Moved accessibility properties off the image and to the button itself
- Adjusted styling for skin tone selector and nearby elements

#### Ticket Link
[GH Issue 18046](https://github.com/mattermost/mattermost-server/issues/18046)
[Jira MM-36733](https://mattermost.atlassian.net/browse/MM-36733)

#### Screenshots
Before/After isn't really possible to show for the main change (making the icon tabbable), however below I have included screenshots showing what the button size, margin, and background adjustments achieve.

(Active) Emoji skin tone selector
|  before  |  after  |
|----|----|
| ![image](https://user-images.githubusercontent.com/9254315/173678836-b89e13e5-e08d-42fd-b209-221b19fa4f6c.png) | ![image](https://user-images.githubusercontent.com/9254315/177218723-7d33e4c4-23bd-48ea-b8ee-547734d2a96d.png)|

(Active) Emoji skin tone selector sub-menu exit
|  before  |  after  |
|----|----|
| ![image](https://user-images.githubusercontent.com/9254315/173678590-ed19dfb9-f199-4d85-940b-fafad3e7e106.png) | ![image](https://user-images.githubusercontent.com/9254315/177218768-12a4bf4a-d4c8-4d26-a6fb-a633d3fb9e5c.png) |

(Active) Emoji skin tone selector sub-menu option
|  before  |  after  |
|----|----|
| ![image](https://user-images.githubusercontent.com/9254315/173678541-ce35179a-e95a-4a1f-95b7-455df5864d1c.png) | ![image](https://user-images.githubusercontent.com/9254315/177218785-abe386e7-673e-48e1-995a-83e003077053.png)|


#### Release Note
Add a release note for each of the following conditions:
```release-note
* Corrects an issue in the Web Application preventing the emoji skin tone picker button from being accessible by tabbing through the emoji picker
* Reverses the order of the skin tone selections in the emoji selector
```

mattermod commented:
Hello @thinkGeist,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

M-ZubairAhmed commented:
@michaelgamble can you please verify is the new CSS changes are appropriate ?

michaelgamble reviewed (changes requested):
A few changes need to made:
- The focus state of the button needs to follow our icon button guide here https://compass.mattermost.com/29be2c109/p/00b099-icon-buttons
- Here is an example of what it could look like in context https://www.figma.com/file/GPTxUI7ACBbbdRfM8Dbtl6/MM-25435-Emoji-Picker-UI-Refresh?node-id=1928%3A62227
<img width="374" alt="image" src="https://user-images.githubusercontent.com/31866506/175993878-31d37f56-9fca-4bc7-909a-be0f32995754.png">
- I noticed that the close icon button in the skin tone selector is not aligned properly - wondering if this is a new button instance or if its leveraging an existing icon button - either way the icon needs to be sized and centered correctly.

Let me know if there are any questions :)

thinkGeist commented:
I'll look into the suggestions above, thanks for the review!

michaelgamble commented:
hmm im aware that this issue might be bigger then this component, i'm going to bring it up to the design team in todays crit meeting to see if we need a larger effort to address this, or if im missing any information as to why their seems to be a larger inconsistency

thinkGeist commented:
@michaelgamble Good call, I'm all for making the adjustments to make this change line up with expectations, I just want to make sure we're making the change in the right place so we don't have to duplicate effort in other parts of the application. 

Let me know what the consensus is and I can look into whatever adjustments remain, thanks again!

thinkGeist commented:
Hey @michaelgamble I've addressed the issues you mentioned above and have updated the screenshots in the ticket description as well. I worked through your list as a sort of "checklist" so I'll drop it again here with what the associated fix was:

1. The search input field should be 284 in width and is currently only ~274 in width
    - Addressed by changing `emoji-picker__text-container` width to 284px
    - Also added an override in _mobile.scss to retain the original 95% width
2. The space between the search field and the skin tone button is currently 14px and it should only be 8px
    - Addressed with margin-right on `emoji-picker__text-container`
3. The space between the search field and the emoji category buttons below seem to be ~12px and it should be 8px
    - Addressed by removing 12px bottom margin from `emoji-picker__search-container` and adding 8px top margin to `emoji-picker__categories`
4. The skintone button's overall width and height seem to be 28x28 when in fact it should be 32x32
    - Addressed by changing the width/height of `skin-tones__icon` from 28 -> 32px
5. The vertical alignment of the hand emoji is not centred vertically (it's too close to the top)
    - Addressed by setting display to flex and align-items to center for `skin-tones__icon`
6. The close icon button vertical position seems to be ~15px from the tab menu above it, it should be 12px
    - Addressed by setting the top margin to 12px on `emoji-picker__search-container`
7. The close icon button horizontal position seems to be ~20px from the left edge, it should be 12px
    - Addressed by setting the left margin to 12px on `emoji-picker__search-container`
8. NOTE: The close icon button size is correctly set differently here at 28x28 (it's using the 'Small - Compact' token as an aesthetic layout choice here - I just wanted to point out the purposeful difference from the skintone button here)
9. The icon here seems to be the incorrect size.. i'm not sure how this was implemented, but it should be using our 16x16 'icon-close' (same icon collection that the emoji category icons are using)
    - I did slightly change how the icon was implemented to leverage the `CloseIcon` from the `@mattermost/compass-icons` package instead of using an `i` element due to some
    suggestions that they should be used going forward in a conversation with michelengelen (here)[https://github.com/mattermost/mattermost-webapp/pull/10559/files/ebd4c35d19ae6409f70f4749b2bfa5f37869a112#r896645602]
    - The size was off in my previous version so I have adjusted the icon size to 16x16 to match your specification above
10. The "Default skin tone" text seems to be 14px from the close button, it should be 15px
    - Addressed by adding a 15px left margin to `skin-tones__close-text`
11. The "Default skin tone" text has a similar issue that is likely being impacted by too much vertical space between itself and the tab menu above (im guessing the fixing the first point in this sublist will fix it)
    - I believe this is fixed with div sizing adjustments
12. The skin tone buttons color progression needs to be reversed
    - Addressed by reversing the order of the `skinsList` constant in `emoji_picker_skin.txt`
13. The skin tone buttons size seems to be 28x28 and should be 32x32
    - Addressed by changing the width/height of `skin-tones__icon` from 28 -> 32px
14. The space between the 32x32 skin tone buttons should be 4px (not sure what its currently set too)
    - Addressed by adding a 4px margin-right to `skin-tones__icon`
15. Just to double check, that the vertical position of the top of the skin tone button to the tab menu above it should be 12px
    - Covered by 12px top margin on `emoji-picker__search-container`
16. Just to double check the space between the last skin tone button and the right edge of the popover should be 12px
    - Covered by 12px left/right margin on `emoji-picker__search-container`

Let me know if there are any further questions, or adjustments that need to be made. Thanks again for taking time to review my changes!

michaelgamble reviewed (changes requested):
@thinkGeist Great work almost nailed it :) In reviewing your last round I found three more items to be addressed:

1. The emoji category icon button row is 10px from the left edge of the popover and should be 12px
2. A weird one - the “skin-tones__close” it seems to be vertically centre aligned with the skin-tones__icon buttons but in the design they are actually all top aligned - Both the close button and the texts alignment can be top-aligned as one object
3. The space between the first skin-tones__icon and the second one seems to be 0px when it should be 4px (the rest seem to be correctly set to 4px
<img width="1011" alt="Pasted Graphic" src="https://user-images.githubusercontent.com/31866506/177353445-3e89b070-fb15-449b-a49b-a0ebe68372db.png">

mattermod commented:
E2E tests not automatically triggered, because the PR is not in a mergeable state. Please update the branch with the base branch and resolve outstanding conflicts.

M-ZubairAhmed commented on components/emoji_picker/components/emoji_picker_skin.tsx:
```diff
@@ -67,29 +71,32 @@ export class EmojiPickerSkin extends React.PureComponent<Props, State> {
-                    className='skin-tones__icon'
+                <button
+                    className='style--none skin-tones__icon'
+                    type='button'
```
any reason we need to specifically mention type button here

thinkGeist replied:
Good eye! I can't see why it would need to be explicitly mentioned here, I'll try removing it and push up the change.

M-ZubairAhmed reviewed (approved):
LGTM except for one small comment

M-ZubairAhmed commented:
/update-branch

M-ZubairAhmed commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/215680

jgilliam17 commented:
Thanks @thinkGeist 
- One test is failing; appears related to this PR - please take a look at the complete Cypress report [here](https://mattermost-cypress-report.s3.amazonaws.com/215680-da16c46-onprem-ent-webapp-pr-10593-mattermost/mm-ee-testda16c46/mochawesome.html). 
test MM-T4438
e2e/cypress/tests/integration/emoji/recently_used_emoji_2_spec.js
- Additionally, while testing on the PR test server, I found the "button, group" announcements confusing, especially for X,  it went something like "you are currently on a button group, to click this button press ..." I expected to hear "you are on a close button" which would give me an indication that clicking on this button would close the group and also a direction to use tab to cycle through options within the group.

thinkGeist commented:
Hey @jgilliam17, thanks for the comments.

I looked into the failing test and found the cause, which I've addressed with my most recent commit. 

As for the screen-reader dialect I'll look into that more and get back to you.

Thanks again!

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/218267
